packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart

```diff
@@ -262,20 +271,178 @@ gmaps.Size? _gmSizeFromIconConfig(List<Object?> iconConfig, int sizeIndex) {
     if (rawIconSize != null) {
       size = gmaps.Size(
-        rawIconSize[0] as num?,
-        rawIconSize[1] as num?,
+        rawIconSize[0]! as double,
+        rawIconSize[1]! as double,
       );
     }
   }
   return size;
 }
 
-// Converts a [BitmapDescriptor] into a [gmaps.Icon] that can be used in Markers.
-gmaps.Icon? _gmIconFromBitmapDescriptor(BitmapDescriptor bitmapDescriptor) {
-  final List<Object?> iconConfig = bitmapDescriptor.toJson() as List<Object?>;
+/// Sets the size of the Google Maps icon with aspect ratio consideration.
+///
+/// This method configures the icon's size based on the provided [width] and
+/// [height]. If both dimensions are null, the size is calculated using the
+/// [imagePixelRatio] based on the actual size of the image fetched from the
+/// [url]. If only one of the dimensions is provided, the other is calculated to
+/// maintain the image's original aspect ratio.
+///
+/// [width] and [height] are optional parameters specifying the width and height
+/// of the icon. If both are null, the dimensions are calculated based on the
+/// image's dimensions divided by [imagePixelRatio].
+///
+/// [imagePixelRatio] is used for adjusting the size calculation when no
+/// dimensions are provided.
+///
+/// [url] is the URL of the image used for the icon. It is required for fetching
+/// the image's actual size when [width] and [height] are not provided.
+///
+/// [cacheKey] is unique key used to identify different images for caching the
+/// result of the size calculation to improve performance.
+///
+/// [icon] is the [gmaps.Icon] object whose size is being set.
+///
+/// This method asserts that [url] is not empty when [width] and [height] are
+/// null, ensuring that a URL is provided for fetching the image.
+Future<void> _setIconSize({
+  required double? width,
+  required double? height,
+  required double imagePixelRatio,
+  required String url,
+  required String cacheKey,
+  required gmaps.Icon icon,
+}) async {
+  if (width != null && height != null) {
+    final gmaps.Size gmapsSize = gmaps.Size(width, height);
+    icon.size = gmapsSize;
+    icon.scaledSize = gmapsSize;
+  } else {
+    assert(
+        url.isNotEmpty, 'URL must not be empty when calculating dimensions.');
+    final Size? bitmapSize = await _getBitmapSize(url, cacheKey);
+
+    // If bitmap size is null, the image is invalid,
+    // and the icon size is not set.
+    if (bitmapSize != null) {
+      double targetWidth = bitmapSize.width;
+      double targetHeight = bitmapSize.height;
+
+      if (width == null && height == null) {
+        // Adjust the target size based on the image pixel ratio.
+        targetWidth /= imagePixelRatio;
+        targetHeight /= imagePixelRatio;
+      } else {
+        final double aspectRatio = bitmapSize.width / bitmapSize.height;
+        targetWidth = width ?? (height ?? bitmapSize.height) * aspectRatio;
+        targetHeight = height ?? (width ?? bitmapSize.width) / aspectRatio;
+      }
```

Comment:
This `_setIconSize` is doing way too many things IMO. Please separate the `_getBitmapSize` call from here, so it's done before calling `_setIconSize`. `_setIconSize` should be super dumb, just receive width, height and icon.

Reply:
Refactored